added 3rd ex need new example

samerrak opened the pull request:
Completed the addition of 3-shot prompting to include the SpaManager Example

samerrak committed:
added 3rd ex need new example

samerrak committed:
finished 3-shot prompting

evanboyd31 commented on backend/resources/n_shot_examples_event_driven.py:
```diff
@@ -235,10 +235,9 @@
+        <tr><td>Level3</td></tr>
         <tr><td>JacuzziPaused</td></tr>
         <tr><td>SaunaOff</td></tr>
         <tr><td>SaunaOn</td></tr>
```
The `SaunaOff` state is in this table twice. `SaunaOn` is a hierarchical state, and hierarchical states should not be in this `states_table`. Remove both of these

```suggestion
```

evanboyd31 commented on backend/resources/n_shot_examples_event_driven.py:
```diff
@@ -248,7 +247,136 @@
+        <tr><td>Pattern Level Up</td></tr>
+        <tr><td>Pattern Level Down</td></tr>
+        <tr><td>Pause Jacuzzi</td></tr>
+        <tr><td>Resume Jacuzzi from Pause</td></tr>
```
The `Resume Jacuzzi from Pause` event is not in the reference solution, so remove this event

```suggestion
```

samerrak replied:
But it is, it's an event where it get's resumed after 2 mins 
Therefore the transition would be from Paused to the History state of On which has the guard as after 2 minutes

evanboyd31 replied:
The `after 2min` isn't a guard in the reference solution. If it were it would be written as `[after 2min]` with the `[]` surrounding it

samerrak replied:
Makes sense, okay committed suggestion

evanboyd31 commented on backend/resources/n_shot_examples_event_driven.py:
````diff
@@ -248,7 +247,136 @@
+         "create_transitions": """```html<table border="1">
+        <tr><th>From State</th><th>To State</th><th>Event</th><th>Guard</th><th>Action</th></tr>
+        <tr><td rowspan="3">JacuzziOff</td><td>JacuzziOn</td><td>Turn Jacuzzi on</td><td>NONE</td><td>NONE</td></tr>
+        </table>```""",
````
When the Event-Driven SMF is using the `state_inspected`, `associated_events`, `event_inspected`, and `create_transitions` examples, it has not yet created hierarchical states. Since the `JacuzziOn` state is a hierarchical state in the reference solution and the LLM will not have information about hierarchical states at this stage in the SMF, we should do a different example here. Something like the following:

```suggestion
        "state_inspected": "Level2",

        "associated_events": "Turn Jacuzzi Off, Pattern Level Up, Pattern Level Down, Set Pattern Type, Pause Jacuzzi",

        "event_inspected": "Pattern Level Up",

         "create_transitions": """```html<table border="1">
        <tr><th>From State</th><th>To State</th><th>Event</th><th>Guard</th><th>Action</th></tr>
        <tr><td rowspan="3">Level2</td><td>Level3</td><td>Pattern Level Up</td><td>NONE</td><td>NONE</td></tr>
        </table>```""",
```

evanboyd31 commented on backend/resources/n_shot_examples_event_driven.py:
```diff
@@ -248,7 +247,136 @@
+        <tr><td>Turn Jacuzzi Off</td></tr>
+        <tr><td>Set Pattern Level 1</td></tr>
+        <tr><td>Set Pattern Level 2</td></tr>
+        <tr><td>Set Pattern Level 3</td></tr>
```
There is a `Set Pattern Type` event defined later in the table, so we can remove these events

```suggestion
```

evanboyd31 commented on backend/resources/n_shot_examples_event_driven.py:
````diff
@@ -248,7 +247,136 @@
+        <tr><td rowspan="3">JacuzziOff</td><td>JacuzziOn</td><td>Turn Jacuzzi on</td><td>NONE</td><td>NONE</td></tr>
+        </table>```""",
+
+        "hierarchical_table":"""```
+        <table border="1">
+        <tr><th>Superstate</th><th>Substate</th></tr>
+        <!-- Jacuzzi section -->
+        <tr><td>SpaManager</td><td>Jacuzzi</td></tr>
+        <tr><td>Jacuzzi</td><td>JacuzziOff</td></tr>
+        <tr><td>Jacuzzi</td><td>JacuzziOn</td></tr>
+        <tr><td>Jacuzzi</td><td>JacuzziPaused</td></tr>
+        <tr><td>JacuzziOn</td><td>Level1</td></tr>
+        <tr><td>JacuzziOn</td><td>Level2</td></tr>
+        <tr><td>JaczziOn</td><td>Level3</td></tr>
+        
+        <!-- Sauna section -->
+        <tr><td>SpaManager</td><td>Sauna</td></tr>
+        <tr><td>Sauna</td><td>SaunaOff</td></tr>
+        <tr><td>Sauna</td><td>SaunaOn</td></tr>
+        <tr><td>SaunaOn</td><td>Heater</td></tr>
+        <tr><td>SaunaOn</td><td>Fan</td></tr>
+        <tr><td>SaunaOn</td><td>Water</td></tr>
+        <tr><td>Heater</td><td>HeaterHeating</td></tr>
+        <tr><td>Heater</td><td>HeaterIdle</td></tr>
+        <tr><td>Fan</td><td>FanOff</td></tr>
+        <tr><td>Fan</td><td>FanOn</td></tr>
+        <tr><td>Water</td><td>WaterIdle</td></tr>
+        </table>```""",
````
It appears that the examples in `backend/resources/n_shot_examples_event_driven.py` do not contain a `parallel_state_table`. 

Since you're already doing a lot with this issue, I think we can just create a new issue to add a `parallel_state_table` entry for each example system. The `EventDrivenParallelRegionsSearchAction` prompts the LLM to perform the following:

```
b. If parallel states are identified:
                - Create an HTML table listing the parallel states and their substates using the following format:
                    ```html
                    <table border="1">
                    <tr>
                        <th>Parallel State</th>
                        <th>Parallel Region</th>
                        <th>Substate</th>
                    </tr>
                    <tr>
                        <td>[Parallel State Name]</td>
                        <td>[Parallel Region Name]</td>
                        <td>[Substate Name]</td>
                    </tr>
                    <!-- Add more rows as needed -->
                    </table>
```

This issue should update each example's `hierarchical_state_table` as well to correctly reflect the parallel states. I can work on this issue

samerrak replied:
created issue

evanboyd31 commented on backend/resources/n_shot_examples_event_driven.py:
```diff
@@ -248,7 +247,136 @@
+        
+        <!-- Heater transitions -->
+        <tr><td>HeaterHeating</td><td>HeaterIdle</td><td>Temperature Exceeds 90°C</td><td>NONE</td><td>NONE</td></tr>
+        <tr><td>HeaterIdle</td><td>HeaterHeating</td><td>Temperature Falls Below 85°C</td><td>NONE</td><td>NONE</td></tr>
```
Missing guards

```suggestion
        <tr><td>HeaterHeating</td><td>HeaterIdle</td><td>Temperature Exceeds 90°C</td><td>[temp >= 90°C]</td><td>NONE</td></tr>
        <tr><td>HeaterIdle</td><td>HeaterHeating</td><td>Temperature Falls Below 85°C</td><td>[temp >= 85°C]</td><td>NONE</td></tr>
```

evanboyd31 commented on backend/resources/n_shot_examples_event_driven.py:
```diff
@@ -248,7 +247,136 @@
+        
+        
+        <!-- Fan transitions -->
+        <tr><td>FanOff</td><td>FanOn</td><td>Humidity Above 40% for 3+ Minutes</td><td>NONE</td><td>NONE</td></tr>
```
Missing guard

```suggestion
        <tr><td>FanOff</td><td>FanOn</td><td>Humidity Above 40% for 3+ Minutes</td><td>[humidity>40% && exceedTime>3min]</td><td>NONE</td></tr>
```

evanboyd31 commented on backend/resources/n_shot_examples_event_driven.py:
```diff
@@ -248,7 +247,136 @@
+        <!-- Jacuzzi transitions -->
+        <tr><td>JacuzziOff</td><td>JacuzziOn</td><td>Turn Jacuzzi On</td><td>NONE</td><td>NONE</td></tr>
+        <tr><td>JacuzziOn</td><td>JacuzziOff</td><td>Turn Jacuzzi Off</td><td>NONE</td><td>NONE</td></tr>
+        <tr><td>JacuzziOn</td><td>JacuzziPaused</td><td>Pause Jacuzzi</td><td>NONE</td><td>NONE</td></tr>
```
When the SMF is at the stage of using the `transitions_table`, it will not have information about hierarchical states. The transitions in this table are later examined by the LLM to generate hierarchical states. These should be modified to:

```suggestion
        <tr><td>JacuzziOff</td><td>Level1</td><td>Turn Jacuzzi On</td><td>NONE</td><td>NONE</td></tr>
        <tr><td>JacuzziOff</td><td>Level2</td><td>Turn Jacuzzi On</td><td>NONE</td><td>NONE</td></tr>
        <tr><td>JacuzziOff</td><td>Level3</td><td>Turn Jacuzzi On</td><td>NONE</td><td>NONE</td></tr>
        <tr><td>Level1</td><td>JacuzziOff</td><td>Turn Jacuzzi Off</td><td>NONE</td><td>NONE</td></tr>
        <tr><td>Level2</td><td>JacuzziOff</td><td>Turn Jacuzzi Off</td><td>NONE</td><td>NONE</td></tr>
        <tr><td>Level3</td><td>JacuzziOff</td><td>Turn Jacuzzi Off</td><td>NONE</td><td>NONE</td></tr>
```

evanboyd31 commented on backend/resources/n_shot_examples_event_driven.py:
```diff
@@ -248,7 +247,136 @@
+    
+        <tr><td>Level1</td><td>JacuzziPaused</td><td>2 Minutes Elapsed (Auto-Pause)</td><td>NONE</td><td>NONE</td></tr>
+        <tr><td>Level2</td><td>JacuzziPaused</td><td>2 Minutes Elapsed (Auto-Pause)</td><td>NONE</td><td>NONE</td></tr>
+        <tr><td>Level3</td><td>JacuzziPaused</td><td>2 Minutes Elapsed (Auto-Pause)</td><td>NONE</td><td>NONE</td></tr>
```
The event should be `Pause Jacuzzi` here

```suggestion
        <tr><td>Level1</td><td>JacuzziPaused</td><td>Pause Jacuzzi</td><td>NONE</td><td>NONE</td></tr>
        <tr><td>Level2</td><td>JacuzziPaused</td><td>Pause Jacuzzi</td><td>NONE</td><td>NONE</td></tr>
        <tr><td>Level3</td><td>JacuzziPaused</td><td>Pause Jacuzzi</td><td>NONE</td><td>NONE</td></tr>
```

evanboyd31 commented on backend/resources/n_shot_examples_event_driven.py:
```diff
@@ -248,7 +247,136 @@
+        <tr><td>Pattern Level Down</td></tr>
+        <tr><td>Pause Jacuzzi</td></tr>
+        <tr><td>Resume Jacuzzi from Pause</td></tr>
+        <tr><td>2 Minutes Elapsed (Auto-Pause)</td></tr>
```
```suggestion
        <tr><td>2 Minutes Elapsed (Auto-Resume)</td></tr>
```

evanboyd31 commented on backend/resources/n_shot_examples_event_driven.py:
````diff
@@ -248,7 +247,136 @@
+
+        "history_state_table": """```html<table border="1"> 
+        <tr><th>From State</th><th>Event</th><th>Guard</th><th>Action</th></tr>
+        <tr><td>JacuzziOn</td><td>Set Pattern Type</td><td>NONE</td><td>NONE</td></tr>
````
```suggestion
        <tr><td>JacuzziOn</td><td>Set Pattern Type</td><td>NONE</td><td>NONE</td></tr>
        <tr><td>JacuzziPaused</td><td>2 Minutes Elapsed (Auto-Resume)</td><td>NONE</td><td>NONE</td></tr>
```

evanboyd31 commented on backend/resources/n_shot_examples_event_driven.py:
```diff
@@ -248,7 +247,136 @@
+        <tr><td>Level2</td><td>JacuzziPaused</td><td>2 Minutes Elapsed (Auto-Pause)</td><td>NONE</td><td>NONE</td></tr>
+        <tr><td>Level3</td><td>JacuzziPaused</td><td>2 Minutes Elapsed (Auto-Pause)</td><td>NONE</td><td>NONE</td></tr>
+        <tr><td>JacuzziPaused</td><td>JacuzziOff</td><td>Turn Jacuzzi Off</td><td>NONE</td><td>NONE</td></tr>
+        
```
Missing transitions from `JacuzziPaused` to `Level1`, `Level2`, and `Level3`. 

```suggestion
        <tr><td>JacuzziPaused</td><td>Level1</td><td>2 Minutes Elapsed (Auto-Resume)</td><td>NONE</td><td>NONE</td></tr>
        <tr><td>JacuzziPaused</td><td>Level2</td><td>2 Minutes Elapsed (Auto-Resume)</td><td>NONE</td><td>NONE</td></tr>
        <tr><td>JacuzziPaused</td><td>Level3</td><td>2 Minutes Elapsed (Auto-Resume)</td><td>NONE</td><td>NONE</td></tr>
```

evanboyd31 commented on backend/resources/n_shot_examples_event_driven.py:
```diff
@@ -248,7 +247,136 @@
+        <tr><td>Temperature Exceeds 90°C</td></tr>
+        <tr><td>Temperature Falls Below 85°C</td></tr>
+        <tr><td>Turn Heater On</td></tr>
+        <tr><td>Heater Goes to Idle</td></tr>
```
These are not events in the reference solution. Remove these

```suggestion
```

samerrak replied:
How is heater goes to idle to not an event?

evanboyd31 replied:
Well I suppose it is, it just isn't explicitly stated in the reference solution. What might be better is to remove the the `Temperature Exceeds 90°C` and `Temperature Falls Below 85°C` events from this table as these events are better suited as guards for the `Heater Goes to Idle` and `Turn Heater On` events, respectively

samerrak replied:
Committed changes please check it out!

evanboyd31 commented on backend/resources/n_shot_examples_event_driven.py:
```diff
@@ -248,7 +247,136 @@
+        <tr><td>Heater Goes to Idle</td></tr>
+        
+        <!-- Fan Events -->
+        <tr><td>Turn Fan On</td></tr>
+        <tr><td>Turn Fan Off</td></tr>
```
These are not events in the reference solution. Remove these

```suggestion
```

samerrak replied:
Events are not transitions in the reference solution you turn the fan on when the guard is met, therefore this is an event. Similar logic for Turn Fan Off

evanboyd31 replied:
Ok makes sense. The `Turn Fan On` and `Turn Fan Off` events can be kept and the `Humidity Above 40% for 3+ Minutes` and `5 Minutes Elapsed (Fan Auto-Off)` events should be removed as these are better suited as guards

samerrak replied:
Committed changes please check it out!

evanboyd31 commented on backend/resources/n_shot_examples_event_driven.py:
```diff
@@ -248,7 +247,136 @@
+        <tr><td>5 Minutes Elapsed (Fan Auto-Off)</td></tr>
+        
+        <!-- Water Events -->
+        <tr><td>Water System to Idle</td></tr>
```
This is not an event in the reference solution. Remove it

```suggestion
```

evanboyd31 commented on backend/resources/n_shot_examples_event_driven.py:
```diff
@@ -248,7 +247,136 @@
+        <tr><td>Water System to Idle</td></tr>
+        <tr><td>Disperse Water</td></tr>
+        <tr><td>Humidity Below 40%</td></tr>
+        <tr><td>15 Minutes Since Last Water Dispersion</td></tr>
```
`Humidity Below 40%` and `15 Minutes Since Last Water Dispersion` are part of the guard condition for the transition from `WaterIdle` to `WaterIdle` in response to the event `DisperseWater`. These two events can be removed from the table

```suggestion
```

samerrak committed:
Update backend/resources/n_shot_examples_event_driven.py

Co-authored-by: Evan Boyd <97850880+evanboyd31@users.noreply.github.com>

samerrak committed:
Update backend/resources/n_shot_examples_event_driven.py

Co-authored-by: Evan Boyd <97850880+evanboyd31@users.noreply.github.com>

samerrak committed:
Update backend/resources/n_shot_examples_event_driven.py

Co-authored-by: Evan Boyd <97850880+evanboyd31@users.noreply.github.com>

samerrak committed:
Update backend/resources/n_shot_examples_event_driven.py

Co-authored-by: Evan Boyd <97850880+evanboyd31@users.noreply.github.com>

samerrak committed:
Update backend/resources/n_shot_examples_event_driven.py

Co-authored-by: Evan Boyd <97850880+evanboyd31@users.noreply.github.com>

samerrak committed:
Update backend/resources/n_shot_examples_event_driven.py

Co-authored-by: Evan Boyd <97850880+evanboyd31@users.noreply.github.com>

samerrak committed:
Update backend/resources/n_shot_examples_event_driven.py

Co-authored-by: Evan Boyd <97850880+evanboyd31@users.noreply.github.com>

samerrak committed:
Update backend/resources/n_shot_examples_event_driven.py

Co-authored-by: Evan Boyd <97850880+evanboyd31@users.noreply.github.com>

samerrak committed:
Update backend/resources/n_shot_examples_event_driven.py

Co-authored-by: Evan Boyd <97850880+evanboyd31@users.noreply.github.com>

samerrak committed:
Update backend/resources/n_shot_examples_event_driven.py

Co-authored-by: Evan Boyd <97850880+evanboyd31@users.noreply.github.com>

samerrak committed:
Update backend/resources/n_shot_examples_event_driven.py

Co-authored-by: Evan Boyd <97850880+evanboyd31@users.noreply.github.com>

samerrak committed:
Update backend/resources/n_shot_examples_event_driven.py

Co-authored-by: Evan Boyd <97850880+evanboyd31@users.noreply.github.com>

samerrak committed:
Update backend/resources/n_shot_examples_event_driven.py

Co-authored-by: Evan Boyd <97850880+evanboyd31@users.noreply.github.com>

samerrak committed:
Update n_shot_examples_event_driven.py

Updated events table and transitions table to reflect changes based on two outstanding suggestions
